server/proxy/proxy.go

```diff
@@ -178,3 +187,67 @@ func (p *Proxy) upstreamForApp(app *apps.App) (upstream.Upstream, error) {
 		return nil, utils.NewInvalidError("not a valid app type: %s", app.AppType)
 	}
 }
+
+func (p *Proxy) CleanUserCallContext(userID string, cc *apps.Context) (*apps.Context, error) {
+	cc = &apps.Context{
+		ContextFromUserAgent: cc.ContextFromUserAgent,
+	}
+
+	var postID, channelID, teamID string
+
+	switch {
+	case cc.PostID != "":
+		postID = cc.PostID
+
+		post, err := p.mm.Post.GetPost(postID)
+		if err != nil {
+			return nil, err
+		}
+
+		channelID = post.ChannelId
+
+		_, err = p.mm.Channel.GetMember(channelID, userID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get channel membership. user=%v channel=%v", userID, channelID)
+		}
+
+		c, err := p.mm.Channel.Get(channelID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get channel. channel=%v", channelID)
+		}
+
+		teamID = c.TeamId
```

Comment:
That seems way to forgiving to me. If a teamID or a chanelID from something the client doesn't have access to is provided, we should just error and not try to correct the input.

Reply:
@hanzei So the logic is currently:

```
if post id is provided:
    - Verify user is a member of the post's channel
        - If they are a member, extract the channel id from the post, and extract the team id from the post's channel. Do not even check what channel id or team id the client provided.
        - If they are not a member, throw an error
else if channel is provided:
    - Verify user is a member of the provided channel
        - If they are a member, extract the team id from the channel. Do not check what team id the client provided.
        - If they are not a member, throw an error
else if team id is provided:
    - Verify user is a member of the provided team
        - If they are a member of the team, use the team id.
        - If they are not a member of the team, throw an error
else: (no ids provided at all)
    - Throw an error
```

Are you suggesting that we instead do something like:

```
if post id is provided:
    - Verify user is a member of the post's channel
        - If they are a member, use this post id and do not use the post's channel id
        - If they are not a member, throw an error
if channel id is provided (no else here):
    - Verify user is a member of the provided channel
        - If they are a member, use this channel and do not use the channel's team id
        - If they are not a member, throw an error
if team id is provided (no else here):
    - Verify user is a member of the provided team
        - If they are a member, use this team
        - If they are not a member, throw an error
```

This second strategy could result in data inconsistencies, but that would be the client's fault. With the first (current) strategy we would be potentially using data that goes against what the client provided. The client should never provide a post id that does not match the channel id it provided but it definitely could (especially if the client is not the webapp or mobile app, i.e. not our code). I don't think we have a use case for providing a channel id that is different than the post's channel id.

Instead of "filling in" the values based on other values (filling in the channel id based on the provided post), are you suggesting we check to see if the channel id provided matches the post and error out if that's not the case?

Reply:
@mickmister Can we assume that every (valid) call request has either
1. Only a `ChannelID`
2. A `PostID` and a `ChannelID`
3. A `PostID` and a `ChannelID` and a `TeamID`
4. A `ChannelID` and a `TeamID`


If so, can we validate these three cases? My main motivation is to reduce the number of RPC calls.

Reply:
@hanzei Currently call requests will always have a channel id and team id (except for the call for fetching bindings due to the PRs below not being merged yet). The "optional" fields are the post id and the root post id, because sometimes the call has to do with a given post, and sometimes the call is just for a given channel and not a specific post.

> If so, can we validate these three cases? My main motivation is to reduce the number of RPC calls.

I think the current implementation in this PR is attempting to minimize the number of RPC calls. I'm not sure how it could be improved. I'm happy to discuss a solution in the MM Apps channel if you'd like.

https://github.com/mattermost/mattermost-webapp/pull/7902
https://github.com/mattermost/mattermost-mobile/pull/5316